packages/react-data-grid/src/cellEvents/Provider.js

```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { EventDispatcher } from './EventDispatcher';
+
+export class CellEventsProvider extends React.Component {
+  static propTypes = {
+    children: PropTypes.isRequired
+  };
+  static childContextTypes = {
+    cellEvents: PropTypes.object.isRequired
+  };
+  constructor() {
+    super();
+    this.cellEvents = new EventDispatcher();
```

Comment:
The pattern looks good, but we probably want to be more generic for the actual implementation and avoid creating new components as much as possible, this should work a little bit like redux, where we initially register a set of modules that can handle events for certain entities, and then we subscribe to events using a generic wrapper around our components. It could work like `combineReducers` in the end

Reply:
This makes me wonder if we should simply use `redux` instead of a custom solution. Another reason is that `context` is [going to be deprecated soon](https://dev.to/kentcdodds/reacts--new-context-api-dpi) and having a 3rd party library taking care of that would be nice. Although, it is not a big change to switch to the new context api.

Reply:
They made changes on the context api so that people don't always need to use third parties for simple jobs.

The main problem with redux will always be the number of things that will re-render when something simple changes, the only way we would be able to stop that from happening would be adding scu in the connected components and that is one of our main design problems atm.

Adding it also means a significant increase in our bundle size and pulling a big library when we only need a fraction of the functionality

Reply:
I am not sure if this is entirely true unless I am missing something. `react-redux` only renders the wrapped component when the state returned by `mapStateToProps` changes so `mapStateToProps` in a way becomes the new `render` and as long this function is simple, there should not be any performance issues. [Here](https://gist.github.com/gaearon/1d19088790e70ac32ea636c025ba424e) is a simplified version of the connect function without any performance optimizations. This is very similar to our custom pub/sub approach and the only difference is we are using `setState` instead of `forceUpdate`. In terms of bundle size, minified `redux` is 5.7 KB and minified `react-redux` is 12.8 KB so it should not increase the bundle size by much (~19 KB in total). Here the [FAQ](https://redux.js.org/docs/faq/Performance.html#how-well-does-redux-scale-in-terms-of-performance-and-architecture) on performance